lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -684,6 +794,77 @@ def _set_unit_ip(self, event):
                 self._charm.model.get_binding(relation).network.bind_address
             )
 
+    def _label_alert_topology(self, rule):
+        """Insert juju topology labels into an alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing Prometheus alert rule with Juju
+            topology labels.
+        """
+        metadata = self._scrape_metadata
+        labels = rule.get("labels", {})
+        labels["juju_model"] = metadata["model"]
+        labels["juju_model_uuid"] = metadata["model_uuid"]
+        labels["juju_application"] = metadata["application"]
+        rule["labels"] = labels
+        return rule
+
+    def _label_alert_expression(self, rule):
+        """Insert juju topology filters into a Prometheus alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing a Prometheus alert rule that filters based
+            on juju topology.
+        """
+        metadata = self._scrape_metadata
+        topology = (
+            'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
+                metadata["model"], metadata["model_uuid"], metadata["application"]
+            )
+        )
+        expr = rule["expr"]  # a rule has to have an "expr"
+        expr = expr.replace("%%juju_topology%%", topology)
+        rule["expr"] = expr
+        return rule
+
+    @property
+    def _labeled_alert_groups(self):
+        """Load alert rules from rule files.
+
+        All rules from files for a consumer charm are loaded into a single
+        group. The generated name of this group includes Juju topology
+        prefixes.
+
+        Returns:
+            a list of Prometheus alert rule groups.
+        """
+        alerts = []
+        for path in Path(ALERT_RULES_PATH).glob("*.rule"):
+            if path.is_file():
+                with path.open() as rule_file:
+                    rules = yaml.safe_load(rule_file)
+                    rule = rules[0]  # each file is list of one rule
+                    rule = self._label_alert_topology(rule)
+                    rule = self._label_alert_expression(rule)
+                    alerts.append(rule)
+        metadata = self._scrape_metadata
+        groups = []
+        if alerts:
+            group = {
+                "name": "{}_{}_{}_alerts".format(
+                    metadata["model"], metadata["model_uuid"], metadata["application"]
```

Comment:
This set of string literals is repeated several times

Reply:
As opposed to defining ALL CAPS global constants and repeating them several times, with the benefit of being able to change them in one place.

Reply:
Not caps necessarily. I would be inclined towards `@dataclass` if there was better interconversion with dicts.

Reply:
Thank you for your suggestion. I would have used things like data classes if I wanted to encapsulate and compose sets of behaviour free information into larger objects. I do think my present use cases warrants this. Perhaps in the future if my code complexity grows I will consider it.